### ISSUE TITLE
Finish docker-ssh end-to-end CI (5 pts)

### ISSUE THREAD

**Issue by pmcharrison** [Dallinger/Dallinger#5094]
As far as we are aware, there are not yet any CI regression tests for the end-to-end `docker-ssh deploy` functionality. We would like to have at least one such test implemented.

Cursor and I developed a proof-of-concept in #8760; the task is then to polish this up and produce something mergable. It is important to emphasise that #8760 is AI-generated and not reviewed, so studying the implementation would be an important part of this task.

Below is a Cursor-generated overview of the required work. I have not reviewed it in detail but have spot-checked various parts. I would add to the below -- feel free to critique the strategy embodied in the existing PR. If it seems flawed, feel free to change direction.

## Summary

Labs do not run PsyNet experiments on the experimenter laptop. Remote debug and live collection go to an Ubuntu box over SSH. Dallinger owns that path (`dallinger docker-ssh`); PsyNet is a thin wrapper (`psynet debug ssh` / `psynet deploy ssh`).

That path has never been exercised in CI. Master only has mocked unit tests. Draft [PR 8760](https://github.com/Dallinger/Dallinger/pull/8760) adds a GitHub Actions job that SSHs into an ephemeral Docker-in-Docker host and runs a real sandbox deploy, plus production fixes the smoke tests found (including a Click `--update` bug PsyNet still works around).

The PR was last green on **21 March 2026**. It is still a **draft**, has **no human review**, is **79 commits behind** master, and **conflicts**. March CI is not evidence it still works. This ticket is: rebase onto current master, keep the runtime fixes, get `docker-ssh-smoke` green on a new SHA, squash, and open it for review. Do not expand scope.

## Motivation

docker-ssh is how this lab ships experiments. When it breaks, people find out on a real server: first deploy, `--update`, TLS, image tags, Postgres users. Unit tests that stub `Executor.run` never see those failures.

A small, gated end-to-end job is the cheapest way to lock the contract:

- `sandbox` can bring an app up
- `apps` / `servers list` see it
- `--update` changes what HTTPS serves without re-running `/launch`
- `destroy` removes it, and a missing app fails clearly

That also unblocks a later PsyNet ticket: once `--update` is a real boolean flag, `psynet debug ssh --update` can stop hard-coding `update=False`.

## Technology overview

Read this before touching the branch. The PR is hard to review without this picture.

### Dallinger vs PsyNet

**Dallinger** is the experiment server: Flask, Postgres, Redis, recruiters (Prolific, MTurk, Hotair, …), dashboard, and the CLIs that put that stack on a machine (local Docker, Heroku, SSH).

**PsyNet** is the experiment-authoring layer used in this lab. Authors write `experiment.py`; they almost never call Dallinger’s SSH commands directly.

| What a lab types | What actually runs | Mode |
|---|---|---|
| `psynet debug ssh --app …` | `dallinger docker-ssh sandbox` | Remote stack, debug/hotair recruitment |
| `psynet deploy ssh --app …` | `dallinger docker-ssh deploy` | Same stack, live recruitment |
| `psynet destroy ssh --app …` | `dallinger docker-ssh destroy` | Tear down one app |
| `psynet apps ssh` / `export ssh` / `test ssh` | Other `docker-ssh` subcommands | Not in scope for this PR |

Implementation: `dallinger/command_line/docker_ssh.py` in Dallinger; PsyNet `ctx.invoke`s those Click commands from `psynet/command_line.py`.

### Registering a server

On the laptop:

1. Passwordless SSH to the box (personal key or a PEM).
2. `server_pem` in `~/.dallingerconfig` pointing at that private key.
3. `dallinger docker-ssh servers add --host my-server.example --user ubuntu`

Dallinger stores host JSON in `~/.dallinger/docker-ssh/hosts/` and will install Docker on the remote if needed. From then on, `--server` is that hostname.

AWS provision (`dallinger ec2 provision`) registers the host for you. Same docker-ssh deploy path afterwards.

### What runs on the remote after a deploy

Two Docker Compose projects, not one.

**Host stack** (`~/dallinger/docker-compose.yml`), shared by every app on that machine:

- **Caddy 2** on 80/443 (TLS + reverse proxy)
- **Postgres 12** (one server, one database per app)
- **Dozzle** (log UI; typically `https://logs.<dns-host>`)

**Per-app stack** (`~/dallinger/<app-id>/docker-compose.yml`):

- `web` — Flask / `dallinger_heroku_web`
- `worker_*` — background jobs
- **Redis** — isolated per app on current master (PR #9374)
- **PgBouncer** — app talks to Postgres through this, not raw 5432
- optional `clock`

The app image is the experiment (demo or PsyNet experiment) plus Dallinger. App containers join the shared `dallinger` network so Caddy can reach `<app>_web:5000`.

HTTPS names are usually `<app-id>.<dns-host>`. Labs with a domain use a wildcard A record. Debug/CI often use **nip.io**, which resolves `<app>.<ip>.nip.io` back to that IP. Caddy config for each app lives in `~/dallinger/caddy.d/<app-id>`; deploy reloads Caddy after writing it.

Dashboard: `https://<app>.<dns>/dashboard`. Experiment pages are the same host.

### How the image gets onto the server

By default Dallinger builds **on the remote**, by setting `DOCKER_HOST=ssh://user@host` and using docker-py. `--local_build` builds on the laptop and pushes to a registry (`docker_image_base_name` in config).

The smoke tests use `--local_build` and `--dns-host nip.io` so GitHub Actions does not need a public registry or a real DNS zone.

Current master tags SSH experiment images with the **per-launch experiment UID**, not a hash of `requirements.txt`. That is newer than PR 8760. After rebase, destroy/update/prune behaviour must still match that tagging.

### Lifecycle: first deploy, `--update`, destroy

**First `sandbox` / `deploy`:**

1. Build or push the image.
2. Ensure the host stack is up (Caddy, Postgres, Dozzle).
3. Write the per-app compose file over SFTP.
4. `CREATE DATABASE` / `CREATE USER` named after the app (unless `--update`).
5. `docker compose up -d` for the app.
6. `dallinger-housekeeper initdb` inside `web`.
7. `POST https://<hostname>/launch` (starts recruitment / experiment setup).
8. Print dashboard URL and log hints.

If the app already exists and `--update` is **not** passed, deploy **aborts**.

**`--update`:** rebuild/restart the app stack, **keep the Postgres database**, **skip `/launch`**. Intended meaning: push new code without wiping participants. The smoke test checks this by changing `templates/instructions/instruct-ready.html` and asserting the HTTPS body changes, and that logs contain `Skipping experiment launch logic because we are in update mode.`

**`destroy`:** remove the Caddy snippet, compose down, and (on current master) remove the compose-pinned experiment image when nothing else still pins it.

### The Click `--update` bug

On master the option is:

```python
click.option("--update", "-u", flag_value="update", default=False, ...)
```

When PsyNet invokes the command in-process, Click can turn that default into the string `"False"`, which is **truthy**. PsyNet therefore hard-codes `update=False` with an explicit comment (`deploy__docker_ssh` / `debug__docker_ssh`). The PR changes the option to `is_flag=True`. Do not drop that fix. Wiring `psynet debug ssh --update` is a **later PsyNet ticket** (old GitLab !519 is obsolete).

### Why a fake SSH target is ugly

CI has no lab VM. The planned job starts a **privileged Ubuntu 24.04 container** on the GitHub runner, installs `openssh-server`, `docker.io`, and `docker-compose-v2` **inside** it, and SSHes to `localhost:<ephemeral-port>`. Nested Docker often needs `--cgroupns=host`, vfs storage, and legacy iptables. The fixture maps host **80 and 443** because Caddy binds those ports. This is why the job is slow (~9 minutes when green) and why March was a long string of infrastructure commits.

## Current solution (master)

| Exists | Gap |
|---|---|
| `tests/test_docker_ssh_apps.py` | Lists/selects apps via a mocked Executor. Never SSHs. |
| `tests/test_docker_ssh_resources.py` | Only warning filters. |
| GitHub `docker` job (`tox -e dockertests`) | Local Docker / image tests, **not** docker-ssh. |
| Production `docker_ssh.py` | Used by every SSH PsyNet run; untested end-to-end. |

Known holes the March work hit: `--update` flag parsing, `host:port` (CI is never port 22), loopback TLS (`HAS_TLS = ssh_host != "localhost"` misses `127.0.0.1`), Paramiko requiring `known_hosts` to exist, Dozzle restart on first deploy when Dozzle is not running, SFTP cwd not `$HOME`, Redis `auto_recruit` read before Redis is up, editable-install image tags.

Master has also moved since March: isolated Redis, `deploy.toml` file planning, UID image tags, config source priority, Python 3.14, newer Actions versions. The PR does not contain those; rebase must preserve them.

## Planned solution (this ticket)

Land a **small, gated smoke suite** on current master, not a general docker-ssh rewrite.

**Fixture** (`dallinger/pytest_docker_ssh.py` on the PR):

- Session-scoped privileged Ubuntu container + nested `dockerd`
- Ed25519 key, `server_pem`, isolated `HOME` so `servers add` does not touch the developer’s real hosts
- Copy of `demos/dlgr/demos/bartlett1932` (canonical Dallinger demo; still has `instruct-ready.html`)
- `DockerSSHServer` helper: `deploy_sandbox`, `update_sandbox`, `destroy_app`, HTTP GET with `verify=False`, remote `docker ps` / logs on failure
- `fresh_docker_ssh_server` resets remote state between tests (today it `docker system prune`s; that makes tests slow — consider pruning less if rebase makes the job too long, but do not add more scenarios)

**Tests** (`tests/test_docker_ssh_integration.py`):

1. sandbox deploy + destroy  
2. `docker-ssh apps` lists the app  
3. `docker-ssh servers list` includes the fixture  
4. destroy of a missing app fails with a clear message  
5. `--update` changes served HTML and skips `/launch`

**Gating:** pytest marker `docker_ssh_smoke` and `--docker-ssh-smoke`. Ordinary pytest/tox must skip these. Regression tests already exist on the PR (`tests/test_pytest_docker_ssh_smoke_selection.py`).

**CI job** `docker-ssh-smoke`: install Dallinger with docker extras, run only that file with the flag, ~45 minute timeout. After rebase, match current workflow style (`checkout@v5`, `setup-python@v6`, Python 3.14 unless 3.13 is required).

**Runtime fixes:** keep the ones the nested topology needs (host:port, loopback TLS, known_hosts, Dozzle guard, SFTP home, Redis fallback, schema `GRANT`, `is_flag=True`, PEP 660 editable path / prerelease image fallback). Optional carve-out: interactive disk-full prune (operator UX, not required for CI). Replace `except Exception: pass` in `get_sftp` with a warning or a real failure.

**Not planned here:** real EC2 in CI, PsyNet Playwright against the remote app, export/stats/clock coverage, replacing DinD with a VM.

## Why PR 8760 is not merge-ready

- Draft, last push 21 March 2026, **0 reviews**
- **Conflicts** with current master, including `docker_ssh.py`, `ci.yml`, `CHANGELOG.md`, `config.py`, `deployment.py`, `docker/tools.py`, `utils.py`, and several test files
- Master since then: Redis isolation, deploy.toml, UID tags, config priority, ResourceWarning filter, Python 3.14, Actions upgrades
- ~57 Cursor-agent commits of “fix the last CI failure” — squash after rebase
- Changelog on the branch is a March snapshot and will conflict; write a **new** Unreleased entry
- PR body does not match `.github/PULL_REQUEST_TEMPLATE.md`

## Work plan

**Read, then rebase.** Read `dallinger/command_line/docker_ssh.py` on **current master** (`_deploy_in_mode`, `build_and_push_image`, `Executor`, destroy) and the March fixture. Branch from `origin/master`. Rebase or cherry-pick the PR. Conflict strategy: keep master’s newer docker-ssh behaviour; re-apply the PR’s host/TLS/Dozzle/`--update`/schema/Redis-fallback fixes on top. If `docker_ssh.py` strategy is unclear, stop and ask before rewriting deploy.

**Prove smoke on the new SHA.** Update the CI job to current Actions versions. Push. Treat nested dockerd, compose plugin, known_hosts, TLS, UID tags, and Redis/container names as first suspects. The fixture already dumps `docker ps` and container logs.

Cheap checks after rebase (no nested Docker):

```bash
python -m pytest tests/test_docker.py tests/test_pytest_docker_ssh_smoke_selection.py tests/test_docker_ssh_apps.py -q
```

Smoke (needs Docker, privileged containers, free 80/443):

```bash
python -m pytest tests/test_docker_ssh_integration.py --docker-ssh-smoke -q
```

Confirm gating: the same file **without** `--docker-ssh-smoke` must skip everything.

**Make it reviewable.** Squash. Changelog. Fill the PR template. Take out of draft. Run `/branch-review` and address findings.

## Out of scope

- PsyNet `--update` / resurrecting GitLab !519  
- Extra smoke cases (export, stats, clock, root-domain, multi-app)  
- Replacing nested Docker with a real VM  
- Dashboard or websocket work  

## Definition of done

- Rebased on current master; GitHub mergeable  
- `docker-ssh-smoke` green on **that** SHA (not March 21)  
- Python 3.11–3.14 and existing `docker` job still green  
- Ordinary pytest does not start nested Docker  
- One Unreleased changelog entry covering the smoke job **and** user-visible docker-ssh fixes  
- PR description matches the Dallinger template; not draft; history squashed  
- `/branch-review` recorded in the PR  

## Reading list (in order)

1. `dallinger/command_line/docker_ssh.py` — `servers add`, `_deploy_in_mode`, `--update`, destroy, `Executor`  
2. `dallinger/docker/ssh_templates/docker-compose-server.yml.j2` and `docker-compose-experiment.yml.j2`  
3. PsyNet `psynet/command_line.py` — `debug__docker_ssh`, `deploy__docker_ssh` (the `update=False` comments)  
4. On the PR branch: `dallinger/pytest_docker_ssh.py`, `tests/test_docker_ssh_integration.py`  
5. Master tests: `tests/test_docker_ssh_apps.py`  
6. Docs: PsyNet `docs/getting_started/running_remotely.rst`, `docs/deploy/ssh_server.rst`  
7. PR: https://github.com/Dallinger/Dallinger/pull/8760  

## Related master PRs that will fight the rebase

- [#9374](https://github.com/Dallinger/Dallinger/pull/9374) Isolate Redis for docker-ssh  
- [#9680](https://github.com/Dallinger/Dallinger/pull/9680) deploy.toml experiment-file planning  
- SSH/Heroku images tagged with launch UID (destroy removes the pinned image)  
- [#9569](https://github.com/Dallinger/Dallinger/pull/9569) config source priority (merge with Redis `auto_recruit` fallback; do not drop either)  
- [#9659](https://github.com/Dallinger/Dallinger/pull/9659) mute docker-py SSH ResourceWarnings
